PXC-3464: Data is not propagated with SET SESSION sql_log_bin = 0

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-3464

Problem:
When sql_log_bin session variable is set to OFF, queries are not
replicated across PXC cluster nodes

Cause:
PXC flow decides if particular statement needs to be replicated basing
on inter alia sql_log_bin value.

Solution:
All statements are replicated. However we are still blocking replication
if it was blocked internally by the server (auxiliary
OPTION_BIN_LOG_INTERNAL_OFF flag introduced)

This work is not based on GCA commit, because currently GCA is too old
and does not contain all necessary changes (code and MTR suite fixes)

kamil-holubicki committed:
PXC-3464: Data is not propagated with
SET SESSION sql_log_bin = 0

https://jira.percona.com/browse/PXC-3464

Problem:
When sql_log_bin session variable is set to OFF, queries are not
replicated across PXC cluster nodes

Cause:
PXC flow decides if particular statement needs to be replicated basing
on inter alia sql_log_bin value.

Solution:
All statements are replicated. However we are still blocking replication
if it was blocked internally by the server (auxiliary
OPTION_BIN_LOG_INTERNAL_OFF flag introduced)

This work is not based on GCA commit, because currently GCA is too old
and does not contain all necessary changes (code and MTR suite fixes)

kamil-holubicki commented:
https://pxc.cd.percona.com/job/pxc-5.7-param/140
No failures caused by the fix.

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_sql_log_bin_zero.test:
```diff
@@ -1,6 +1,6 @@
-# Test SET SESSION sql_log_bin = 0 . We expect that unlogged updates will not be replicated
-# to the slave and that there will be no assertions in the process.
+# Test SET SESSION sql_log_bin = 0 . We expect that even with this settings
+# repliation across PXC nodes will take place.
```
s/repliation/replication

venkatesh-prasad-v commented on sql/bootstrap.cc:
```diff
@@ -278,8 +287,12 @@ static void handle_bootstrap_impl(THD *thd)
     */
     if (last_query_source == QUERY_SOURCE_COMPILED &&
-        thd->variables.option_bits & OPTION_BIN_LOG)
+        thd->variables.option_bits & OPTION_BIN_LOG) {
```
I think this should also come under `#ifdef WITH_WSREP`

venkatesh-prasad-v commented on sql/bootstrap.cc:
```diff
@@ -119,6 +119,9 @@ static void handle_bootstrap_impl(THD *thd)
   Query_command_iterator query_iter(bootstrap_query);
   bool has_binlog_option= thd->variables.option_bits & OPTION_BIN_LOG;
+#ifdef WITH_WSREP
+  ulonglong option_bits_save = thd->variables.option_bits;
```
Multiple occurences: MySQL coding style - no space before assignment operator.

venkatesh-prasad-v commented on sql/handler.cc:
```diff
@@ -8453,14 +8453,20 @@ static bool check_table_binlog_row_based(THD *thd, TABLE *table)
-	  /* applier and replayer should not binlog */
-          ((WSREP_EMULATE_BINLOG(thd) && (thd->wsrep_exec_mode != REPL_RECV)) ||
-           mysql_bin_log.is_open()));
+          !(thd->variables.option_bits & OPTION_BIN_LOG_INTERNAL_OFF) &&
+            (((WSREP_EMULATE_BINLOG(thd) && (thd->wsrep_exec_mode != REPL_RECV))
+            ||
+            (((WSREP(thd) || thd->variables.option_bits & OPTION_BIN_LOG))
+              &&
+              mysql_bin_log.is_open()))));
```
Can we add a comment explaining the need of `OPTION_BIN_LOG_INTERNAL_OFF` in https://github.com/percona/percona-xtradb-cluster/blob/649422b36e3f411fc8ffaa630dc8596cb028d837/sql/handler.cc#L8431-L8441

kamil-holubicki committed:
PXC-3464: Data is not propagated with
SET SESSION sql_log_bin = 0

https://jira.percona.com/browse/PXC-3464

Review comments addressed.

venkatesh-prasad-v commented on sql/handler.cc:
```diff
@@ -8454,6 +8454,13 @@ static bool check_table_binlog_row_based(THD *thd, TABLE *table)
 
 #ifdef WITH_WSREP
+ /*
+    In addtion to above conditions, PXC allows replication if:
```
s/addtion/addition/

kamil-holubicki replied:
fixed

venkatesh-prasad-v reviewed (approved):
Thank you for working on this and for addressing the review comments.
Patch looks great now. So, it is approved from my side.

kamil-holubicki committed:
PXC-3464: Data is not propagated with
SET SESSION sql_log_bin = 0

https://jira.percona.com/browse/PXC-3464

Use wait_condition instead of assert to prevent replication lag failures
